backend/src/rate_limiter/mod.rs

```diff
@@ -0,0 +1,179 @@
+
+        RLKey { key }
+    }
+
+    pub(crate) fn serialize(&self) -> &[u8] {
+        &self.key
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Allowance {
+    remaining: u64,
+    valid_until: DateTime<Utc>,
+}
+
+impl Allowance {
+    pub(crate) fn new(config: &RLConfig) -> Self {
+        Allowance {
+            remaining: config.max_requests,
+            valid_until: Utc::now() + config.time_window,
+        }
+    }
+
+    fn reset(&mut self, config: &RLConfig) {
+        self.remaining = config.max_requests;
+        self.valid_until = Utc::now() + config.time_window;
+    }
+
+    fn allowed(&mut self, config: &RLConfig) -> bool {
+        // Check if the time window has passed
+        if self.valid_until < Utc::now() {
+            self.reset(config);
+        }
+
+        if self.remaining == 0 {
+            false
+        } else {
+            self.remaining -= 1;
+            true
+        }
+    }
+}
+
+#[async_trait]
+pub(crate) trait StorageProvider {
+    async fn get(&self, key: &RLKey) -> Option<Allowance>;
+    async fn set(&self, key: RLKey, allowance: Allowance);
+}
+
+pub(crate) struct RateLimiter<S: StorageProvider> {
+    config: RLConfig,
+    storage: S,
+}
+
+impl<S: StorageProvider> RateLimiter<S> {
+    pub(crate) fn new(config: RLConfig, storage: S) -> Self {
+        RateLimiter { config, storage }
+    }
+
+    pub(crate) async fn allowed(&self, key: RLKey) -> bool {
+        let mut allowance = self
+            .storage
+            .get(&key)
+            .await
+            .unwrap_or_else(|| Allowance::new(&self.config));
+
+        if allowance.allowed(&self.config) {
+            self.storage.set(key, allowance.clone()).await;
+            true
+        } else {
+            false
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use chrono::TimeDelta;
+    use tokio::sync::Mutex;
+    use tonic::async_trait;
+
+    use crate::rate_limiter::{RLConfig, RateLimiter};
+
+    use super::{Allowance, RLKey, StorageProvider};
+
+    #[derive(Default)]
+    pub struct InMemoryStorage {
+        data: Mutex<HashMap<Vec<u8>, Allowance>>,
+    }
+
+    impl InMemoryStorage {
+        pub fn new() -> Self {
+            InMemoryStorage {
+                data: Mutex::new(HashMap::new()),
+            }
+        }
+    }
+
+    #[async_trait]
+    impl StorageProvider for InMemoryStorage {
+        async fn get(&self, key: &RLKey) -> Option<Allowance> {
+            self.data.lock().await.get(key.serialize()).cloned()
+        }
+
+        async fn set(&self, key: RLKey, allowance: Allowance) {
+            self.data
+                .lock()
+                .await
+                .insert(key.serialize().to_owned(), allowance);
+        }
+    }
+
+    #[tokio::test]
+    async fn test_rate_limiter() {
+        let config = RLConfig {
+            max_requests: 5,
+            time_window: TimeDelta::seconds(1),
```

Comment:
Minor: maybe we can make it 100 ms? No need to run a test for longer than 1 sec.